lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -0,0 +1,151 @@
+from http import HTTPStatus
+from logging import Logger
+
+from flask.json import jsonify
+from flask.wrappers import Request, Response
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    ApiBase,
+    API_OPERATION,
+    JSON,
+    Parameter,
+    ParamType,
+    Schema,
+    SchemaError,
+)
+from pbench.server.database.database import Database
+from pbench.server.database.models.datasets import (
+    Dataset,
+    MetadataError,
+)
+
+
+class DatasetsList(ApiBase):
+    """
+    API class to list datasets based on PostgreSQL metadata
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("name", ParamType.STRING),
+                Parameter("owner", ParamType.USER),
+                Parameter("access", ParamType.ACCESS),
+                Parameter("controller", ParamType.STRING),
+                Parameter("start", ParamType.DATE),
+                Parameter("end", ParamType.DATE),
+                Parameter(
+                    "metadata",
+                    ParamType.LIST,
+                    element_type=ParamType.KEYWORD,
+                    keywords=ApiBase.METADATA,
+                    required=False,
+                ),
+            ),
+            role=API_OPERATION.READ,
+        )
+
+    def _get(self, json_data: JSON, request: Request) -> Response:
+        """
+        Get a list of datasets matching a set of criteria.
+
+        NOTE: This does not rely on a JSON payload to identify the dataset and
+        desired metadata keys; instead we rely on URI query parameters.
+
+        TODO: The dispatcher is currently merging JSON request body and Flask
+        URI keywords ... it probably makes sense to also merge request query
+        parameters and then validate them all together; although it might make
+        more sense to validate each against separate schemas.
```

Comment:
Yeah, I had similar thoughts; in principle one name could mean different things in the context of a Flask URI template, a query parameter, and a JSON parameter. (Not that I'm suggesting this would be a good idea.)

That suggests having separate `Schema` objects for each context, which is clear but a bit bulky, or possibly enhancing `Parameter` with an ENUM to specify the context. (Possibly a list, although again I'm kinda against encouraging anyone to cross the streams.)

Then again, being able to specify, say, `metadata` either in a JSON body or as a query parameter has a certain level of appeal. In that case, we *would* want a single consistent meaning, and, yeah, we'd have to define a priority when merging. E.g., does `GET ... ?metadata=dataset.created` with `{"metadata": "dataset.owner"}` as a body return *both*, or does the JSON body supersede the query parameter, or vice versa? On balance, I don't like the idea of dealing with this sort of semantic; I'd rather keep them completely separate namespaces and not deal with that complication.